src/basename.rs

```diff
@@ -20,15 +20,88 @@ pub fn basename<'a>(path: &'a str, ext: &str) -> &'a str {
       left = last_slash_pos + 1;
     }
   }
+  &path[left..left + ext_end(&bytes[left..right], ext)]
+}
+
+fn ext_end(slice: &[u8], ext: &str) -> usize {
+  if ext.len() >= slice.len() || slice == b"." || slice == b".." {
+    return slice.len();
+  }
   let ext_bytes = ext.as_bytes();
+  // TODO: Implement other ext patterns (see the ignored test below).
   if ext_bytes == b".*" {
-    if let Some(dot_pos) = find_last_dot_pos(&bytes[left..right]) {
-      right = left + dot_pos;
+    if let Some(dot_pos) = find_last_dot_pos(slice) {
+      return dot_pos;
     }
-  } else if bytes[left..right].ends_with(ext_bytes) {
-    right -= ext_bytes.len();
+  } else if slice.ends_with(ext_bytes) {
+    return slice.len() - ext_bytes.len();
   }
-  &path[left..right]
+  slice.len()
+}
+
+// TODO: Un-ignore once the functionality has been implemented.
```

Comment:
Just so you know Ruby's `File.basename` method explicitly only uses `.*` for splat and nothing else.  So any other implementation would just be for our benefit in Rust and not a Ruby compatible version.

Reply:
That's not the case, I've tested all the examples in the tests case in `irb` in MRI 2.5.0.

See also the Ruby implementation:
https://github.com/ruby/ruby/blob/5a53cbe314203ce63630030110c51e7650d0fc0a/file.c#L4222

But it does seem to be undocumented.

Reply:
You can search through Ruby's test suite and the only thing they tests against is `.*` https://github.com/ruby/ruby/blob/3527c05a8f4e189772cdac17f166bd9626c24661/spec/ruby/core/file/basename_spec.rb I'm not sure where I got this information from that long ago but it's rooted in Ruby's practice if not their rule.

Reply:
That's true, but code may still rely on this undocumented behaviour.
I've figured out an easy way to match Ruby exactly here, will send in a moment.

Reply:
I've updated the PR and it should now match even that undocumented Ruby behaviour.